# INFO 6105 Assignment 4 

## Jian Min 001491533

# Question1.1

What is the Probability Distribution for each F1 driver to win the Singaporean Grand Prix? 

What is the Probability Distribution for each F1 driver to win both the Singaporean and Russian Grand Prix? 

What is the probability for Mercedes to win both races? 

What is the probability for Mercedes to win at least one race?

In [1]:
def p(event, space): 
    """The probability of an event, given a sample space of equiprobable outcomes. 
    event: a collection of outcomes, or a predicate that is true of outcomes in the event. 
    space: a set of outcomes or a probability distribution of {outcome: frequency} pairs."""
    if is_predicate(event):
        event = such_that(event, space)
    if isinstance(space, ProbDist):
        return sum(space[o] for o in space if o in event)
    else:
        return Fraction(len(event & space), len(space))

is_predicate = callable

def such_that(predicate, space): 
    """The outcomes in the sample pace for which the predicate is true.
    If space is a set, return a subset {outcome,...} with outcomes where predicate(element) is true;
    if space is a ProbDist, return a ProbDist {outcome: frequency,...} with outcomes where predicate(element) is true."""
    if isinstance(space, ProbDist):
        return ProbDist({o:space[o] for o in space if predicate(o)})
    else:
        return {o for o in space if predicate(o)}

In [2]:
class ProbDist(dict):
    """A Probability Distribution; an {outcome: probability} mapping."""
    def __init__(self, mapping=(), **kwargs):
        self.update(mapping, **kwargs)
        # Make probabilities sum to 1.0; assert no negative probabilities
        total = sum(self.values())
        for outcome in self:
            self[outcome] = self[outcome] / total
            assert self[outcome] >= 0

## 1.1.1 Probability Distribution for each F1 driver to win the Singaporean Grand Prix

import data and get the probability distribution in one race

In [3]:
SGP = ProbDist(
    LH = 413,
    VB = 326,
    MV = 278,
    CL = 264,
    SV = 240,
    CS = 96,
    PG = 95,
    AA = 92,
    DR = 54,
    SP = 52,
    LN = 49,
    KR = 43,
    DK = 37,
    NH = 37,
    LS = 21,
    KM = 20,
    AG = 14,
    RG = 8,
    RK = 1,
    GR = 0
)
SGP

## 1.1.2 Probability Distribution for each F1 driver to win both the Singaporean and Russian Grand Prix

the two race is just the same, we can get the probabilities by taking its square

In [4]:
def f(x): return x ** 2.0
SRGP = {k: f(v) for k, v in SGP.items()}
SRGP

or use leverage

In [5]:
def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b]
                    for a in A
                    for b in B})
SRGP = joint(SGP, SGP, ' ')

In [6]:
winners = SRGP.keys()
probs = SRGP.values()

#check the first winner is the same as the second one
double_winners = {x for x in winners if x[0] == x[3] and x[1] == x[4]}
probs = [SRGP[k] for k in double_winners]
double_winners_probs = list(zip(double_winners, probs))
print('\n'.join([str(outcome) for outcome in double_winners_probs]))
sum(probs)

('SP SP', 0.000590444580312691)
('AG AG', 4.279849768538737e-05)
('LS LS', 9.629661979212154e-05)
('PG PG', 0.001970696130666433)
('DR DR', 0.0006367368329111712)
('SV SV', 0.012577517687134243)
('MV MV', 0.016875709668966718)
('LN LN', 0.0005242815966459952)
('KR KR', 0.000403747052144292)
('VB VB', 0.023206393571490954)
('CS CS', 0.0020124028299414788)
('CL CL', 0.015218796401432434)
('RG RG', 1.3975019652371384e-05)
('NH NH', 0.0002989344047515066)
('DK DK', 0.0002989344047515066)
('AA AA', 0.0018481963490261156)
('LH LH', 0.037245392610708354)
('KM KM', 8.734387282732114e-05)
('RK RK', 2.1835968206830287e-07)
('GR GR', 0.0)


0.11394881649052317

## 1.1.3 Probability for Mercedes to win both races

meet the Mercedes drivers

In [7]:
Mercedes = ['LH','VB']

write a function to check the keys if they meet the situation: Mercedes wins both races

In [8]:
def are_both_Mercedes(keys):
    count = 0;
    for p in Mercedes:
        count += keys.count(p)
    if count == 2 : return True
    else: return False;

get the keys via the function

In [9]:
winners = SRGP.keys()
probs = SRGP.values()
Mercedes_win_both = {x for x in winners if are_both_Mercedes(x)}
Mercedes_win_both

{'LH LH', 'LH VB', 'VB LH', 'VB VB'}

get the values(probabilities) and sum them

In [10]:
Mercedes_probs_both = [SRGP[k] for k in Mercedes_win_both]
Mercedes_wins_both_prob = sum(Mercedes_probs_both)
print('Probability for Mercedes to win both races: ' + str(Mercedes_wins_both_prob))

Probability for Mercedes to win both races: 0.11925080793082363


4 Probability for Mercedes to win at least one race

write a function to check the keys meets the situation: Mercedes winx both races

In [11]:
def at_least_one_Mercedes(keys):
    flag = False;
    for p in Mercedes:
        if keys.count(p) > 0 : flag = True
    return flag;

get the keys via the function

In [12]:
Mercedes_wins_at_least_one = {x for x in winners if at_least_one_Mercedes(x)}
len(Mercedes_wins_at_least_one)

76

get the values(probabilities) and sum them

In [13]:
Mercedes_probs_at_least_one = [SRGP[k] for k in Mercedes_wins_at_least_one]
Mercedes_wins_at_least_one_prob = sum(Mercedes_probs_at_least_one)
print('Probability for Mercedes to win at least one race: ' + str(Mercedes_wins_at_least_one_prob))

Probability for Mercedes to win at least one race: 0.5714033976766527


# Question1.2

If Mercedes wins the first race, what is the probability that Mercedes wins the next one? 

If Mercedes wins at least one of these two races, what is the probability Mercedes wins both races? 

How about Ferrari, Red Bull, and Renault?

## 1.2.1 If Mercedes wins the first race, the probability that Mercedes wins the next one

Because their are many teams, I change these functions that can take the space of each team

In [14]:
# the team wins the first race
def wins_first(keys,space):
    flag = False;
    for p in space:
        if keys[0]+keys[1] == p : flag = True
    return flag;
# the team wins the next race
def wins_next(keys,space):
    flag = False;
    for p in space:
        if keys[3]+keys[4] == p : flag = True
    return flag;
# the team wins at least one of the races
def at_least_one(keys,space):
    flag = False;
    for p in space:
        if keys.count(p) > 0 : flag = True
    return flag;
# the team wins both of the races
def wins_both(keys,space):
    count = 0;
    for p in space:
        count += keys.count(p)
    if count == 2 : return True
    else: return False;

the getSpace function can rebuild the SRGP dict to the situation we need 

In [15]:
#Get new sample space based on condition
def getSpace(f,space):
    new_dict = {}
    denominator = 0;
    for x in SRGP:
        if(f(x,space)):
            #update keys
            new_dict.update({x : SRGP[x]})
            denominator += SRGP[x]
    #change value based the sample space
    for i in new_dict:
        new_dict[i] = new_dict[i]/denominator
    return new_dict

now we set the situation of Mercedes wins the first race 

In [16]:
#MX is the sample space of Mercedes wins first race 
MX = getSpace(wins_first,Mercedes)
len(MX)

40

the size of space matchs, there should be 2*20 kinds of results(Mercedes wins the first one, anyone wins the next one)

Then contiune computing the sum of the probabilities in the MX space which belongs to the situation that Mercedes wins the next one

In [17]:
winners = MX.keys()
probs = MX.values()
Mercedes_wins_next_if_MX = {x for x in winners if wins_next(x,Mercedes)}
Mercedes_probs_wins_next_if_MX = [MX[k] for k in Mercedes_wins_next_if_MX if wins_next(k,Mercedes)]
Mercedes_wins_both_if_MX_prob = sum(Mercedes_probs_wins_next_if_MX)
print('If Mercedes wins the first race, the probability that Mercedes wins the next one: \n' +
      str(Mercedes_wins_both_if_MX_prob))

If Mercedes wins the first race, the probability that Mercedes wins the next one: 
0.34532710280373835


So if Mercedes wins the first race, the probability that Mercedes wins the next one is up to 34%

## Verification of 1.2.1

To verify the result above is correct, we need to get all the probabilities that any one wins the next race if Mercedes wins the first race, and add them all together to see if it equals 1

In [18]:
Mercedes = ['LH','VB']
Ferrari = ['CL','SV']
RedBull = ['MV','AA']
Renault = ['DR','NH']
McLaren = ['CS','LN']
AlfaRomeo = ['KR','AG']
Scuderia = ['PG','DK']
RacingPoint = ['SP','LS']
HAAS = ['KM','RG']
WILLIAMS = ['RK','GR']

I write a fucntion to make the verification efficient

In [19]:
def autoGenerate(condition,result,condition_space,result_space):
    new_space = getSpace(condition,condition_space)
    winners = new_space.keys()
    probs = new_space.values()
    result_winners = {x for x in winners if result(x,result_space)}
    result_probs = [new_space[k] for k in result_winners if result(k,result_space)]
    result_sum_prob = sum(result_probs)
    return result_sum_prob;

In [20]:
#Mercedes
#don't forget this one
Mercedes_prob = Mercedes_wins_both_if_MX_prob
print('Mercedes_prob: ' + str(Mercedes_prob))
#Ferrari
Ferrari_prob = autoGenerate(wins_first,wins_next,Mercedes,Ferrari)
print('Ferrari_prob: ' + str(Ferrari_prob))
#RedBull
RedBull_prob = autoGenerate(wins_first,wins_next,Mercedes,RedBull)
print('RedBull_prob: ' + str(RedBull_prob))
#Renault 
Renault_prob = autoGenerate(wins_first,wins_next,Mercedes,Renault)
print('Renault_prob: ' + str(Renault_prob))
#McLaren
McLaren_prob = autoGenerate(wins_first,wins_next,Mercedes,McLaren)
print('McLaren_prob: ' + str(McLaren_prob))
#AlfaRomeo
AlfaRomeo_prob = autoGenerate(wins_first,wins_next,Mercedes,AlfaRomeo)
print('AlfaRomeo_prob: ' + str(AlfaRomeo_prob))
#Scuderia
Scuderia_prob = autoGenerate(wins_first,wins_next,Mercedes,Scuderia)
print('Scuderia_prob: ' + str(Scuderia_prob))
#RacingPoint
RacingPoint_prob = autoGenerate(wins_first,wins_next,Mercedes,RacingPoint)
print('RacingPoint_prob: ' + str(RacingPoint_prob))
#HAAS
HAAS_prob = autoGenerate(wins_first,wins_next,Mercedes,HAAS)
print('HAAS_prob: ' + str(HAAS_prob))
#WILLIAMS
WILLIAMS_prob = autoGenerate(wins_first,wins_next,Mercedes,WILLIAMS)
print('WILLIAMS_prob: ' + str(WILLIAMS_prob))

#SUM
allProbs = [Mercedes_prob,
            Ferrari_prob,
            RedBull_prob,
            Renault_prob,
            McLaren_prob,
            AlfaRomeo_prob,
            Scuderia_prob,
            RacingPoint_prob,
            HAAS_prob,
            WILLIAMS_prob]
sumProb = sum(allProbs)
print("Add them all and get: " + str(sumProb))


Mercedes_prob: 0.34532710280373835
Ferrari_prob: 0.23551401869158883
RedBull_prob: 0.1728971962616823
Renault_prob: 0.04252336448598132
McLaren_prob: 0.0677570093457944
AlfaRomeo_prob: 0.026635514018691596
Scuderia_prob: 0.06168224299065421
RacingPoint_prob: 0.03411214953271029
HAAS_prob: 0.013084112149532713
WILLIAMS_prob: 0.0004672897196261683
Add them all and get: 1.0000000000000002


if the sum value is (almost) 1, then the function works in the correct way

## 1.2.2 If Mercedes wins at least one of these two races, the probability Mercedes wins both races

we set the situation of Mercedes wins the first race

In [21]:
#M1 is the sample space of Mercedes wins at least one of these two races
M1 = getSpace(at_least_one,Mercedes)
len(M1)

76

the size of space matchs, there should be 

2*18 (Mercedes wins the first one, one of the rest wins the next one) +

18*2 (Mercedes wins the next one, one of the rest wins the first one) +

2*2（Mercedes wins the both）

which is 76 kinds of results 

In [22]:
winners = M1.keys()
probs = M1.values()
Mercedes_wins_both_if_M1 = {x for x in winners if wins_both(x,Mercedes)}
Mercedes_probs_wins_both_if_M1 = [M1[k] for k in Mercedes_wins_both_if_M1 if wins_both(k,Mercedes)]
Mercedes_wins_both_if_M1_prob = sum(Mercedes_probs_wins_both_if_M1)
print('If Mercedes wins at least one of these two races, the probability Mercedes wins both races \n' +
      str(Mercedes_wins_both_if_M1_prob))

If Mercedes wins at least one of these two races, the probability Mercedes wins both races 
0.20869810787913018


## Verification of 1.2.2

To verify the result above is correct, we need to get all the probabilities that any one wins the other race if Mercedes wins one races, and add them all together to see if it equals 1

With the last verification's functions, the results are directly

In [23]:
#Mercedes
#don't forget this one
Mercedes_prob = Mercedes_wins_both_if_M1_prob
print('Mercedes_prob: ' + str(Mercedes_prob))
#Ferrari
Ferrari_prob = autoGenerate(at_least_one,at_least_one,Mercedes,Ferrari)
print('Ferrari_prob: ' + str(Ferrari_prob))
#RedBull
RedBull_prob = autoGenerate(at_least_one,at_least_one,Mercedes,RedBull)
print('RedBull_prob: ' + str(RedBull_prob))
#Renault 
Renault_prob = autoGenerate(at_least_one,at_least_one,Mercedes,Renault)
print('Renault_prob: ' + str(Renault_prob))
#McLaren
McLaren_prob = autoGenerate(at_least_one,at_least_one,Mercedes,McLaren)
print('McLaren_prob: ' + str(McLaren_prob))
#AlfaRomeo
AlfaRomeo_prob = autoGenerate(at_least_one,at_least_one,Mercedes,AlfaRomeo)
print('AlfaRomeo_prob: ' + str(AlfaRomeo_prob))
#Scuderia
Scuderia_prob = autoGenerate(at_least_one,at_least_one,Mercedes,Scuderia)
print('Scuderia_prob: ' + str(Scuderia_prob))
#RacingPoint
RacingPoint_prob = autoGenerate(at_least_one,at_least_one,Mercedes,RacingPoint)
print('RacingPoint_prob: ' + str(RacingPoint_prob))
#HAAS
HAAS_prob = autoGenerate(at_least_one,at_least_one,Mercedes,HAAS)
print('HAAS_prob: ' + str(HAAS_prob))
#WILLIAMS
WILLIAMS_prob = autoGenerate(at_least_one,at_least_one,Mercedes,WILLIAMS)
print('WILLIAMS_prob: ' + str(WILLIAMS_prob))

#SUM
allProbs = [Mercedes_prob,
            Ferrari_prob,
            RedBull_prob,
            Renault_prob,
            McLaren_prob,
            AlfaRomeo_prob,
            Scuderia_prob,
            RacingPoint_prob,
            HAAS_prob,
            WILLIAMS_prob]
sumProb = sum(allProbs)
print("Add them all and get: " + str(sumProb))

Mercedes_prob: 0.20869810787913018
Ferrari_prob: 0.2846653487715334
RedBull_prob: 0.20898051397910197
Renault_prob: 0.0513979101948602
McLaren_prob: 0.08189776899181021
AlfaRomeo_prob: 0.03219429539678057
Scuderia_prob: 0.07455521039254448
RacingPoint_prob: 0.041231290595876866
HAAS_prob: 0.015814741598418523
WILLIAMS_prob: 0.0005648121999435188
Add them all and get: 0.9999999999999998


if the sum value is (almost) 1, then the function works in the correct way

## 1.2.3 Other teams

Since the scope of teams is not the same in BlackBoard and notebook, I take all the teams that mentioned

In [24]:
Ferrari = ['CL','SV']
RedBull = ['MV','AA']
Renault = ['DR','NH']
McLaren = ['CS','LN']
AlfaRomeo = ['KR','AG']

I wrote this function after I finished all the part, and I tried it in the Ferrari part, the endless variable names makes me crazy :( 

In [25]:
def autoGenerate(condition,result,condition_space,result_space):
    new_space = getSpace(condition,condition_space)
    winners = new_space.keys()
    probs = new_space.values()
    result_winners = {x for x in winners if result(x,result_space)}
    result_probs = [new_space[k] for k in result_winners if result(k,result_space)]
    result_sum_prob = sum(result_probs)
    return result_sum_prob;

## Ferrari:

1 If Ferrari wins the first race, the probability that Ferrari wins the next one: 

In [26]:
#FX is the sample space of Ferrari wins first race 
FX = getSpace(wins_first,Ferrari)
len(FX)

40

In [27]:
winners = FX.keys()
probs = FX.values()
Ferrari_wins_next_if_FX = {x for x in winners if wins_next(x,Ferrari)}
Ferrari_probs_wins_next_if_FX = [FX[k] for k in Ferrari_wins_next_if_FX if wins_next(k,Ferrari)]
Ferrari_wins_both_if_FX_prob = sum(Ferrari_probs_wins_next_if_FX)
print('If Ferrari wins the first race, the probability that Ferrari wins the next one: \n' +
      str(Ferrari_wins_both_if_FX_prob))

If Ferrari wins the first race, the probability that Ferrari wins the next one: 
0.23551401869158872


auto generater test:

    condtion:wins_first
    
    result:wins_next
    
    space:Ferrari

In [28]:
print(autoGenerate(wins_first,wins_next,Ferrari,Ferrari))

0.23551401869158872


2 If Ferrari wins at least one of these two races, the probability Ferrari wins both races

In [29]:
#F1 is the sample space of Ferrari wins at least one of these two races
F1 = getSpace(at_least_one,Ferrari)
len(F1)

76

In [30]:
winners = F1.keys()
probs = F1.values()
Ferrari_wins_both_if_F1 = {x for x in winners if wins_both(x,Ferrari)}
Ferrari_probs_wins_both_if_F1 = [F1[k] for k in Ferrari_wins_both_if_F1 if wins_both(k,Ferrari)]
Ferrari_wins_both_if_F1_prob = sum(Ferrari_probs_wins_both_if_F1)
print('If Ferrari wins at least one of these two races, the probability Ferrari wins both races \n' +
      str(Ferrari_wins_both_if_F1_prob))

If Ferrari wins at least one of these two races, the probability Ferrari wins both races 
0.1334745762711864


## RedBull

1 If RedBull wins the first race, the probability that RedBull wins the next one:

In [31]:
#RbX is the sample space of RedBull wins first race 
RbX = getSpace(wins_first,RedBull)
len(RbX)

40

In [32]:
winners = RbX.keys()
probs = RbX.values()
RedBull_wins_next_if_RbX = {x for x in winners if wins_next(x,RedBull)}
RedBull_probs_wins_next_if_RbX = [RbX[k] for k in RedBull_wins_next_if_RbX if wins_next(k,RedBull)]
RedBull_wins_both_if_RbX_prob = sum(RedBull_probs_wins_next_if_RbX)
print('If RedBull wins the first race, the probability that RedBull wins the next one: \n' +
      str(RedBull_wins_both_if_RbX_prob))

If RedBull wins the first race, the probability that RedBull wins the next one: 
0.17289719626168232


2 If RedBull wins at least one of these two races, the probability RedBull wins both races

In [33]:
#Rb1 is the sample space of RedBull wins at least one of these two races
Rb1 = getSpace(at_least_one,RedBull)
len(Rb1)

76

In [34]:
winners = Rb1.keys()
probs = Rb1.values()
RedBull_wins_both_if_Rb1 = {x for x in winners if wins_both(x,RedBull)}
RedBull_probs_wins_both_if_Rb1 = [Rb1[k] for k in RedBull_wins_both_if_Rb1 if wins_both(k,RedBull)]
RedBull_wins_both_if_Rb1_prob = sum(RedBull_probs_wins_both_if_Rb1)
print('If RedBull wins at least one of these two races, the probability RedBull wins both races \n' +
      str(RedBull_wins_both_if_Rb1_prob))

If RedBull wins at least one of these two races, the probability RedBull wins both races 
0.09462915601023018


## Renault

1 If Renault wins the first race, the probability that Renault wins the next one:

In [35]:
#RX is the sample space of Renault wins first race 
RX = getSpace(wins_first,Renault)
len(RX)

40

In [36]:
winners = RX.keys()
probs = RX.values()
Renault_wins_next_if_RX = {x for x in winners if wins_next(x,Renault)}
Renault_probs_wins_next_if_RX = [RX[k] for k in Renault_wins_next_if_RX if wins_next(k,Renault)]
Renault_wins_both_if_RX_prob = sum(Renault_probs_wins_next_if_RX)
print('If Renault wins the first race, the probability that Renault wins the next one: \n' +
      str(Renault_wins_both_if_RX_prob))

If Renault wins the first race, the probability that Renault wins the next one: 
0.04252336448598132


2 If Renault wins at least one of these two races, the probability Renault wins both races

In [37]:
#R1 is the sample space of Renault wins at least one of these two races
R1 = getSpace(at_least_one,Renault )
len(R1)

76

In [38]:
winners = R1.keys()
probs = R1.values()
Renault_wins_both_if_R1 = {x for x in winners if wins_both(x,Renault)}
Renault_probs_wins_both_if_R1 = [R1[k] for k in Renault_wins_both_if_R1 if wins_both(k,Renault)]
Renault_wins_both_if_R1_prob = sum(Renault_probs_wins_both_if_R1)
print('If Renault wins at least one of these two races, the probability Renault wins both races \n' +
      str(Renault_wins_both_if_R1_prob))

If Renault wins at least one of these two races, the probability Renault wins both races 
0.021723561709238483


## McLaren

1 If McLaren wins the first race, the probability that McLaren wins the next one:

In [39]:
#McX is the sample space of Renault wins first race 
McX = getSpace(wins_first,McLaren)
len(McX)

40

In [40]:
winners = McX.keys()
probs = McX.values()
McLaren_wins_next_if_McX = {x for x in winners if wins_next(x,McLaren)}
McLaren_probs_wins_next_if_McX = [McX[k] for k in McLaren_wins_next_if_McX if wins_next(k,McLaren)]
McLaren_wins_both_if_McX_prob = sum(McLaren_probs_wins_next_if_McX)
print('If McLaren wins the first race, the probability that McLaren wins the next one: \n' +
      str(McLaren_wins_both_if_McX_prob))

If McLaren wins the first race, the probability that McLaren wins the next one: 
0.06775700934579439


2 If McLaren wins at least one of these two races, the probability McLaren wins both races

In [41]:
#Mc1 is the sample space of McLaren wins at least one of these two races
Mc1 = getSpace(at_least_one,McLaren)
len(Mc1)

76

In [42]:
winners = Mc1.keys()
probs = Mc1.values()
McLaren_wins_both_if_Mc1 = {x for x in winners if wins_both(x,McLaren)}
McLaren_probs_wins_both_if_Mc1 = [Mc1[k] for k in McLaren_wins_both_if_Mc1 if wins_both(k,McLaren)]
McLaren_wins_both_if_Mc1_prob = sum(McLaren_probs_wins_both_if_Mc1)
print('If McLaren wins at least one of these two races, the probability McLaren wins both races \n' +
      str(McLaren_wins_both_if_Mc1_prob))

If McLaren wins at least one of these two races, the probability McLaren wins both races 
0.03506650544135429


## AlfaRomeo

1 If AlfaRomeo wins the first race, the probability that AlfaRomeo wins the next one:

In [43]:
#AX is the sample space of Renault wins first race 
AX = getSpace(wins_first,AlfaRomeo)
len(AX)

40

In [44]:
winners = AX.keys()
probs = AX.values()
AlfaRomeo_wins_next_if_AX = {x for x in winners if wins_next(x,AlfaRomeo)}
AlfaRomeo_probs_wins_next_if_AX = [AX[k] for k in AlfaRomeo_wins_next_if_AX if wins_next(k,AlfaRomeo)]
AlfaRomeo_wins_both_if_AX_prob = sum(AlfaRomeo_probs_wins_next_if_AX)
print('If AlfaRomeo wins the first race, the probability that AlfaRomeo wins the next one: \n' +
      str(AlfaRomeo_wins_both_if_AX_prob))

If AlfaRomeo wins the first race, the probability that AlfaRomeo wins the next one: 
0.026635514018691592


2 If AlfaRomeo wins at least one of these two races, the probability AlfaRomeo wins both races

In [45]:
#A1 is the sample space of AlfaRomeo wins at least one of these two races
A1 = getSpace(at_least_one,AlfaRomeo)
len(A1)

76

In [46]:
winners = A1.keys()
probs = A1.values()
AlfaRomeo_wins_both_if_A1 = {x for x in winners if wins_both(x,AlfaRomeo)}
AlfaRomeo_probs_wins_both_if_A1 = [A1[k] for k in AlfaRomeo_wins_both_if_A1 if wins_both(k,AlfaRomeo)]
AlfaRomeo_wins_both_if_A1_prob = sum(AlfaRomeo_probs_wins_both_if_A1)
print('If AlfaRomeo wins at least one of these two races, the probability AlfaRomeo wins both races \n' +
      str(AlfaRomeo_wins_both_if_A1_prob))

If AlfaRomeo wins at least one of these two races, the probability AlfaRomeo wins both races 
0.013497513615912862


# Question1.3

Mercedes wins at least one of these two races on a rainy day. 

What is the probability Mercedes wins both races, assuming races can be held on either rainy, sunny, cloudy, snowy or foggy days?

First, we create weather space and each type of weather has the same probability

In [47]:
weather = ProbDist(
    R = 1,#Rainy
    S = 1,#Sunny
    C = 1,#Cloudy
    Sn = 1,#Snowy
    F = 1 #Foggy
)

get joint probability of the two race

In [48]:
SGP_weather = joint(SGP, weather, '')
SRGP_weather = joint(SGP_weather , SGP_weather,' ')
len(SRGP_weather)

10000

In [49]:
len(SGP_weather)

100

Then we need some new functions:

In [50]:
#check the situation that any team win at least one of the races on a rainy day
def at_least_one_R(keys,space):
    flag = False;
    for p in space:
        if keys.count(p +'R') > 0 : flag = True
    return flag;
Mercedes = ['LH','VB']
#get sample space from SRGP_weather
def getSpaceWeather(f,space):
    new_dict = {}
    denominator = 0;
    for x in SRGP_weather:
        if(f(x,space)):
            #update keys
            new_dict.update({x : SRGP_weather[x]})
            denominator += SRGP_weather[x]
    #change value based the sample space
    for i in new_dict:
        new_dict[i] = new_dict[i]/denominator
    return new_dict

In [51]:
#M1_R is the sample space of Mercedes wins at least one of these two races on a rainy day
M1_R = getSpaceWeather(at_least_one_R,Mercedes)
len(M1_R)

396

And we can check the size:

(2 * 1) * (18 * 5) (Mercedes wins the first race on a rainy day, the one of rest team wins the next one on any kinds day) +

(18 * 5) * (2 * 1) (Mercedes wins the next race on a rainy day, the one of rest team wins the first one on any kinds day) +

(2 * 1) * (2 * 4) (Mercedes wins the both races, only first race is a rainy day) +

(2 * 4) * (2 * 1) (Mercedes wins the both races, only next race is a rainy day) +

(2 * 1) * (2 * 1) (Mercedes wins the both races, and they are both rainy day) +

which is 396 kinds results

so it's correct

And question is "what is probability Mercedes wins both races" in the space M1_R

(It's lucky my first wins_both function won't be impact by the character order changed in the key, so I don't need to renew it.)

In [52]:
winners = M1_R.keys()
probs = M1_R.values()
Mercedes_wins_both_if_M1_R = {x for x in winners if wins_both(x,Mercedes)}
Mercedes_prob_wins_both_if_M1_R = [M1_R[k] for k in Mercedes_wins_both_if_M1_R if wins_both(k,Mercedes)]
Mercedes_wins_both_if_M1_R_prob = sum(Mercedes_prob_wins_both_if_M1_R)
print('If Mercedes wins at least one of these two races on a rainy day, the probability that Mercedes wins both races: \n' +
      str(Mercedes_wins_both_if_M1_R_prob))

If Mercedes wins at least one of these two races on a rainy day, the probability that Mercedes wins both races: 
0.3219108465224339


32% is quite good

After seeing the result, I guess the drivers of Mercedes may want to win in a rainy day :)